ckanext/s3filestore/uploader.py

```diff
@@ -132,18 +135,47 @@ def clear_key(self, filepath):
         except Exception as e:
             raise e
 
-    def get_signed_url_to_key(self, key_path, expiredin=3600):
-        # Small workaround to manage downloading of large files
-        # We are using redirect to resource public URL
+    def get_signed_url_to_key(self, key_path, expiredin=3600, cache_window=30):
+        '''Generates a pre-signed URL giving access to an S3 object.
+
+        To enable effective caching, the timestamp used in the signature
+        is rounded to the previous half hour (by default),
+        so that a consistent URL will be returned for 30 minutes.
+        Eg a URL generated at 12:34 might have a signature timestamp
+        of 12:30.
+
+        Cacheability can be controlled by setting the 'expiredin' value,
+        which controls how long a signed URL is valid, and the
+        'cache_window' value, which controls how far back the timestamp
+        will be truncated. Note that 'expiredin' uses seconds, while
+        'cache_window' uses minutes. The expiry must be longer than the
+        truncation (not equal); otherwise, a URL may expire before a new
+        one is available. The default expiry is one hour.
+
+        If a download_proxy is configured, then the URL will be
+        generated using the true S3 host, and then the hostname will be
+        rewritten afterward. Note that the Host header is part of a
+        version 4 signature, so the resulting request, as it stands,
+        will fail signature verification; the download_proxy server must
+        be configured to set the Host header back to the true value when
+        forwarding the request (CloudFront does this automatically).
+        '''
         client = self.get_s3_client()
 
         # check whether the object exists in S3
         client.head_object(Bucket=self.bucket_name, Key=key_path)
 
-        url = client.generate_presigned_url(ClientMethod='get_object',
-                                            Params={'Bucket': self.bucket_name,
-                                                    'Key': key_path},
-                                            ExpiresIn=expiredin)
+        signature_time = datetime.datetime.now(tzlocal.get_localzone())
+        cache_minute = int((math.floor(signature_time.minute / cache_window))) * cache_window
+        signature_time = signature_time.replace(
+            minute=cache_minute,
+            second=0,
+            microsecond=0)
+        with freezegun.freeze_time(signature_time):
+            url = client.generate_presigned_url(ClientMethod='get_object',
+                                                Params={'Bucket': self.bucket_name,
+                                                        'Key': key_path},
+                                                ExpiresIn=expiredin)
```

Comment:
do we unfreeze time?
Can you add a test of some sort that timestamp is altered and after use its back to normal time?
see Project: S4   Author: MichaelAquilina   File: test_s3.py 
https://www.programcreek.com/python/example/105984/freezegun.freeze_time

Reply:
That's what the `with` keyword is for. 'freezegun' supports the "context manager" approach, see https://pypi.org/project/freezegun/ and https://docs.python.org/2.5/whatsnew/pep-343.html